### PR TITLE
Documentation: Update docs tool to omit private, experimental, unstable APIs

### LINKED ISSUE #15148: Documentation: Omit unstable / experimental functions from data documentation
Data documentation includes references to unstable and experimental functions. These should be omitted.

https://wordpress.org/gutenberg/handbook/designers-developers/developers/data/data-core-block-editor/#__unstableislastblockchangeignored

Unstable and experimental APIs are not to be documented:

>An experimental or unstable API is named as such to communicate instability of a function whose interface is not yet finalized. Aside from references within the code, **these APIs should neither be documented nor mentioned in any CHANGELOG**. They should effectively be considered to not exist from an external perspective. In most cases, they should only be exposed to satisfy requirements between packages maintained in this repository.

https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-and-unstable-apis

The `@wordpress/docgen` tool is already configured to omit these, but the data documentation generation is a separate process:

https://github.com/WordPress/gutenberg/tree/master/docs/tool

### PR DESCRIPTION
**Pull request by aduth**
Closes #15148

This pull request seeks to update the docs tool to omit selectors and actions which are either named as unstable or experimental, or which include a `@private` JSDoc tag.

- https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-and-unstable-apis
- http://usejsdoc.org/tags-private.html

**Implementation notes:**

There is certainly some overlap / potential for inconsistency between the documentation output by this tool, and that of `@wordpress/docgen`. These should probably be reconciled, but it seemed a larger effort than the immediate goal of excluding APIs not intended to be documented.

**Testing instructions:**

Verify there are no unstable, experimental, or private APIs documented.

Verify there are no local changes after running `npm run docs:build` in this branch.

### CODE REVIEW

File: docs/tool/parser.js [hunk 1]
```diff
@@ -25,6 +25,29 @@ function isDocumentedNamedExport( node ) {
...
+ * Returns true if the given exported declaration name is considered stable for
+ * documentation, or false otherwise.
+ *
+ * @see https://github.com/WordPress/gutenberg/blob/master/docs/contributors/coding-guidelines.md#experimental-and-unstable-apis
+ *
+ * @param {string} name Name to test.
+ *
+ * @return {boolean} Whether the provided name describes a stable API.
+ */
+const isStableExportName = ( name ) => ! /^__(unstable|experimental)/.test( name );
```
Comment (gziolo) [inline docs/tool/parser.js]: Can we make it case insensitive? I'd like it to cover component names as well:

`__UnstableComponent` or `__ExperimentalComponent`.

Comment (aduth) [reply docs/tool/parser.js]: > Can we make it case insensitive? I'd like it to cover component names as well:
> 
> `__UnstableComponent` or `__ExperimentalComponent`.

AFAIK (and what I'd recommend), it's still capitalized as `__experimentalComponent`.

Example:

https://github.com/WordPress/gutenberg/blob/da09d36f20a26e3ba4c89ad8c1186a11284741b7/packages/data/src/index.js#L16

**Comment by aduth** [comment]
> There is certainly some overlap / potential for inconsistency between the documentation output by this tool, and that of `@wordpress/docgen`. These should probably be reconciled, but it seemed a larger effort than the immediate goal of excluding APIs not intended to be documented.

#15200 is a step toward this.

File: docs/tool/parser.js (hunk 1 above)
Comment (gziolo) [reply docs/tool/parser.js]: Okay, let's follow this pattern 👍

**Review by gziolo — APPROVED** [review]
Let's move forward with it and focus our efforts on moving `selectors` and `actions` under control of `docgen` 👍

**Pull request merged**